File: kevoree-tools/org.kevoree.tools.aether.framework/src/main/scala/org/kevoree/tools/aether/framework/NodeTypeBootstrapHelper.scala
/**
 * Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3, 29 June 2007;
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.gnu.org/licenses/lgpl-3.0.txt
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kevoree.tools.aether.framework

import org.kevoree.api.service.core.handler.KevoreeModelHandlerService
import org.kevoree.api.service.core.script.KevScriptEngineFactory
import org.kevoree.api.Bootstraper
import org.kevoree.api.service.core.classloading.KevoreeClassLoaderHandler
import org.kevoree.framework.{AbstractGroupType, KevoreeGeneratorHelper, AbstractNodeType}
import org.kevoree.framework.aspects.KevoreeAspects._
import java.io.File
import org.kevoree.{DeployUnit, KevoreeFactory, GroupType, ContainerRoot}
import org.kevoree.extra.jcl.KevoreeJarClassLoader

/**
 * User: ffouquet
 * Date: 10/08/11
 * Time: 12:01
 */

class NodeTypeBootstrapHelper extends Bootstraper with KCLBootstrap {

  val classLoaderHandler = new JCLContextHandler

  def bootstrapNodeType(model: ContainerRoot, destNodeName: String, mservice: KevoreeModelHandlerService, kevsEngineFactory: KevScriptEngineFactory): Option[org.kevoree.api.NodeType] = {
    //JCLContextHandler.clear()


    //LOCATE NODE
    val nodeOption = model.getNodes.find(node => node.getName == destNodeName)
    nodeOption match {
      case Some(node) => {
        val nodeTypeDeployUnitList = node.getTypeDefinition.getDeployUnits.toList
        if (nodeTypeDeployUnitList.size > 0) {
          val classLoader = installNodeType(node.getTypeDefinition.asInstanceOf[org.kevoree.NodeType])
          if (classLoader.isDefined) {
            val clazz: Class[_] = classLoader.get.loadClass(node.getTypeDefinition.getBean)
            val nodeType = clazz.newInstance.asInstanceOf[AbstractNodeType]
            //ADD INSTANCE DICTIONARY
            val dictionary: java.util.HashMap[String, AnyRef] = new java.util.HashMap[String, AnyRef]
            node.getTypeDefinition.getDictionaryType.map {
              dictionaryType =>
                dictionaryType.getDefaultValues.foreach {
                  dv =>
                    dictionary.put(dv.getAttribute.getName, dv.getValue)
                }
            }
            node.getDictionary.map {
              dictionaryModel =>
                dictionaryModel.getValues.foreach {
                  v =>
                    dictionary.put(v.getAttribute.getName, v.getValue)
                }
            }
            //   dictionary.put(Constants.KEVOREE_PROPERTY_OSGI_BUNDLE, bundleContext.getBundle)
            nodeType.setDictionary(dictionary)
            nodeType.setNodeName(destNodeName)
            //INJECT SERVICE HANDLER
            //  val sr = bundleContext.getServiceReference(classOf[KevoreeModelHandlerService].getName)
            // if (sr != null) {
            //   val s = bundleContext.getService(sr).asInstanceOf[KevoreeModelHandlerService]
            nodeType.setModelService(mservice)
            nodeType.setKevScriptEngineFactory(kevsEngineFactory)
            nodeType.setBootStrapperService(this)
            // }
            //nodeType.push(destNodeName, model, bundle.getBundleContext)
            Some(nodeType)
          } else {
            None
          }
          //KevoreeDeployManager.addMapping(KevoreeOSGiBundle(node.getTypeDefinition.getName, node.getTypeDefinition.getClass.getName, lastBundleID))
        } else {
          logger.error("NodeType deploy unit not found , have you forgotten to merge nodetype library ?");
          None
        }
      }
      case None => logger.error("Node not found using name " + destNodeName); None
    }
  }


  /* Bootstrap node type bundle in local environment */
  private def installNodeType(nodeType: org.kevoree.NodeType): Option[ClassLoader] = {
    val superTypeBootStrap = nodeType.getSuperTypes
      .forall(superType => installNodeType(superType.asInstanceOf[org.kevoree.NodeType]).isDefined)
    if (superTypeBootStrap) {
      var kcl: ClassLoader = null
      nodeType.getDeployUnits.forall(ct => {
        val dpRes = ct.getRequiredLibs.forall(tp => {
          val idp = installDeployUnit(tp)
          idp.isDefined
        })
        /*
        groupKCL.foreach(gcl =>
          groupKCL.filter(g => g != gcl).foreach(
            gcl_in => {
              gcl.addWeakClassLoader(gcl_in)
            }
          )
        )*/

        val kcl_opt = installDeployUnit(ct)
        kcl_opt match {
          case Some(k) => {
            kcl = k
          }
          case _ =>
        }
        kcl_opt.isDefined && dpRes
      })
      Some(kcl) //TODO
    } else {
      logger.error("Super type of " + nodeType.getName + " was not completely installed")
      None
    }
  }

  def getKevoreeClassLoaderHandler: KevoreeClassLoaderHandler = {
    classLoaderHandler
  }


  def bootstrapGroupType(model: ContainerRoot, destGroupName: String, mservice: KevoreeModelHandlerService): Option[AbstractGroupType] = {
    //LOCATE NODE
    val optgroup = model.getGroups.find(group => group.getName == destGroupName)
    optgroup match {
      case Some(group) => {
        val groupTypeDeployUnitList = group.getTypeDefinition.getDeployUnits.toList
        if (groupTypeDeployUnitList.size > 0) {


          val kcl = installGroupTyp(group.getTypeDefinition.asInstanceOf[GroupType])
          if (kcl.isDefined) {
            val activatorPackage = KevoreeGeneratorHelper.getTypeDefinitionGeneratedPackage(group.getTypeDefinition, "JavaSENode")
            val activatorName = group.getTypeDefinition.getName + "Activator"



            val clazz: Class[_] = kcl.get.loadClass(activatorPackage + "." + activatorName)

            val groupActivator = clazz.newInstance.asInstanceOf[org.kevoree.framework.osgi.KevoreeGroupActivator]
            val groupType = groupActivator.callFactory()

            //ADD INSTANCE DICTIONARY
            val dictionary: java.util.HashMap[String, AnyRef] = new java.util.HashMap[String, AnyRef]

            group.getTypeDefinition.getDictionaryType.map {
              dictionaryType =>
                dictionaryType.getDefaultValues.foreach {
                  dv =>
                    dictionary.put(dv.getAttribute.getName, dv.getValue)
                }
            }
            group.getDictionary.map {
              dictionaryModel =>
                dictionaryModel.getValues.foreach {
                  v =>
                    dictionary.put(v.getAttribute.getName, v.getValue)
                }
            }
            //    dictionary.put(Constants.KEVOREE_PROPERTY_OSGI_BUNDLE, bundleContext.getBundle)
            groupType.getDictionary().putAll(dictionary)
            groupType.setName(destGroupName)
            //INJECT SERVICE HANDLER
            //  val sr = bundleContext.getServiceReference(classOf[KevoreeModelHandlerService].getName)
            // if (sr != null) {
            //    val s = bundleContext.getService(sr).asInstanceOf[KevoreeModelHandlerService]
            groupType.setModelService(mservice)
            groupType.asInstanceOf[AbstractGroupType].setBootStrapperService(this)
            //  }
            Some(groupType)
          } else {
            None
          }
        } else {
          logger.error("NodeType deploy unit not found , have you forgotten to merge nodetype library ?");
          None
        }
      }
      case None => logger.error("Node not found using name " + destGroupName); None
    }
  }

  private def installGroupTyp(groupType: GroupType): Option[ClassLoader] = {
    val superTypeBootStrap = groupType.getSuperTypes.forall(superType => installGroupTyp(superType.asInstanceOf[GroupType]).isDefined)
    if (superTypeBootStrap) {
      //FAKE NODE TODO
      val fakeNode = KevoreeFactory.createContainerNode
      groupType.eContainer.asInstanceOf[ContainerRoot].getTypeDefinitions.find(td => td.getName == "JavaSENode").map {
        javaseTD =>
          fakeNode.setTypeDefinition(javaseTD)
      }
      val ct = groupType.foundRelevantDeployUnit(fakeNode)
      if (ct != null) {
        var kcl: ClassLoader = null
        val dpRes = ct.getRequiredLibs.forall {
          tp => installDeployUnit(tp).isDefined
        }
        val kcl_opt = installDeployUnit(ct)
        kcl_opt match {
          case Some(k) => {
            kcl = k
          }
          case _ =>
        }
        kcl_opt.isDefined && dpRes
        Some(kcl) //TODO
        //groupType.getDeployUnits.forall(ct => {
        /*
        logger.debug("require lib for " + ct.getUnitName + "->" + ct.getRequiredLibs.size)
        ct.getRequiredLibs.forall {
          tp => installDeployUnit(tp, bundleContext)
        } && installDeployUnit(ct, bundleContext)
        //})*/
      } else {
        None
      }
    } else {
      None
    }
  }

  def resolveArtifact(artId: String, groupId: String, version: String, repos: List[String]): File = AetherUtil.resolveMavenArtifact(artId,groupId,version,repos)

  def resolveKevoreeArtifact(artId: String, groupId: String, version: String): File = AetherUtil.resolveKevoreeArtifact(artId,groupId,version)

  def resolveDeployUnit(du: DeployUnit): File = AetherUtil.resolveDeployUnit(du)

  def close {
    classLoaderHandler.clear()
    classLoaderHandler.stop()
  }

  def clear {
    classLoaderHandler.clear()
  }

  def registerManuallyDeployUnit(artefactID: String,groupID:String,version:String,kcl:KevoreeJarClassLoader){
    val du = KevoreeFactory.createDeployUnit
    du.setUnitName(artefactID)
    du.setGroupName(groupID)
    du.setVersion(version)
    classLoaderHandler.manuallyAddToCache(du,kcl)
  }

}

File: kevoree-tools/org.kevoree.tools.aether.framework/src/main/java/org/kevoree/tools/aether/framework/NodeTypeBootstrapHelper.kt
package org.kevoree.tools.aether.framework

import org.kevoree.api.service.core.handler.KevoreeModelHandlerService
import org.kevoree.api.service.core.script.KevScriptEngineFactory
import org.kevoree.api.Bootstraper
import org.kevoree.api.service.core.classloading.KevoreeClassLoaderHandler
import org.kevoree.framework.*
import java.io.File
import org.kevoree.*
import org.kevoree.kcl.KevoreeJarClassLoader
import java.util
import org.kevoree.impl.DefaultKevoreeFactory
import java.util.ArrayList
import org.kevoree.framework.kaspects.TypeDefinitionAspect
import org.kevoree.log.Log
import java.util.List
import org.kevoree.kcl.Klassloader

/**
 * User: ffouquet
 * Date: 10/08/11
 * Time: 12:01
 */

open class NodeTypeBootstrapHelper : Bootstraper, KCLBootstrap {

    override fun resolveKevoreeArtifact(artId: String, groupId: String, version: String): File? {
        val l = ArrayList<String>()
        if(version.contains("SNAPSHOT") || version.contains("LATEST")){
            l.add("http://oss.sonatype.org/content/groups/public/")
        } else {
            l.add("http://repo1.maven.org/maven2")
        }
        return resolveArtifact(artId, groupId, version, l)
    }

    val kevoreeFactory: KevoreeFactory = DefaultKevoreeFactory()
    var _kevoreeLogService: org.kevoree.api.service.core.logging.KevoreeLogService? = null

    override fun getKevoreeLogService(): org.kevoree.api.service.core.logging.KevoreeLogService {
        return _kevoreeLogService!!
    }
    override fun setKevoreeLogService(ls: org.kevoree.api.service.core.logging.KevoreeLogService) {
        _kevoreeLogService = ls
    }

    var classLoaderHandler = JCLContextHandler()

    override fun bootstrapNodeType(model: ContainerRoot, destNodeName: String, mservice: KevoreeModelHandlerService, kevsEngineFactory: KevScriptEngineFactory): org.kevoree.api.NodeType? {
        //LOCATE NODE
        val nodeOption = model.findNodesByID(destNodeName)
        if(nodeOption != null){
            val nodeTypeDeployUnit = nodeOption.typeDefinition!!.deployUnit
            if (nodeTypeDeployUnit!=null) {
                val classLoader = installNodeType(nodeOption.typeDefinition as org.kevoree.NodeType)
                if (classLoader != null) {
                    val clazz = classLoader.loadClass(nodeOption!!.typeDefinition!!.bean)
                    val nodeType = clazz!!.newInstance() as AbstractNodeType
                    //ADD INSTANCE DICTIONARY
                    val dictionary = java.util.HashMap<String, Any>()
                    val dictionaryType = nodeOption!!.typeDefinition!!.dictionaryType
                    if (dictionaryType != null) {
                        for(dv in dictionaryType.defaultValues) {
                            dictionary.put(dv!!.attribute!!.name!!, dv.value!!)
                        }
                    }

                    val dictionaryModel = nodeOption.dictionary

                    if (dictionaryModel != null) {
                        for (v in dictionaryModel.values) {
                            dictionary.put(v!!.attribute!!.name!!, v.value!!)
                        }
                    }
                    nodeType.setDictionary(dictionary)
                    nodeType.setNodeName(destNodeName)
                    //INJECT SERVICE HANDLER
                    nodeType.setModelService(mservice)
                    nodeType.setKevScriptEngineFactory(kevsEngineFactory)
                    nodeType.setBootStrapperService(this)
                    return nodeType
                } else {
                    return null
                }
            } else {
                Log.error("NodeType deploy unit not found , have you forgotten to merge nodetype library ?")
                return null
            }
        } else {
            Log.error("Node not found using name " + destNodeName);
            return null
        }
    }


    /* Bootstrap node type bundle in local environment */
    private fun installNodeType(nodeType: org.kevoree.NodeType): Klassloader? {
        val superKCLs = ArrayList<Klassloader>()
        val superTypeBootStrap = nodeType.superTypes.all {
            superType ->
            val superKCL = installNodeType(superType as org.kevoree.NodeType)
            if (superKCL != null) {
                superKCLs.add(superKCL)
            }
            superKCL != null
        }
        if (superTypeBootStrap) {
            var kcl: Klassloader? = null
                val kcl_opt = recursivelyInstallDeployUnit(nodeType.deployUnit!!)
                if(kcl_opt != null){
                    kcl = kcl_opt
                }
                kcl_opt != null
            if (kcl == null) {
                Log.error("Unable to install the Node Type, maybe some libs cannot be installed")
            } else {
                superKCLs.forEach {
                    superKCL ->
                    (kcl as KevoreeJarClassLoader).addSubClassLoader(superKCL)
                }
            }
            return kcl //TODO
        } else {
            Log.error("Super type of " + nodeType.name + " was not completely installed")
            return null
        }
    }


    override fun getKevoreeClassLoaderHandler(): KevoreeClassLoaderHandler {
        return classLoaderHandler
    }


    override fun resolveDeployUnit(du: DeployUnit): File? = AetherUtil.resolveDeployUnit(du)

    override fun close() {
        classLoaderHandler.clear()
        classLoaderHandler.stop()
    }

    override fun clear() {
        classLoaderHandler.clear()
    }

    fun registerManuallyDeployUnit(artefactID: String, groupID: String, version: String, kcl: KevoreeJarClassLoader) {
        val du = kevoreeFactory.createDeployUnit()
        du.name = artefactID
        du.groupName = groupID
        du.version = version
        classLoaderHandler.manuallyAddToCache(du, kcl)
    }

    override fun resolveArtifact(artId: String, groupId: String, version: String, extension: String, repos: jet.List<String>): File? {
        return AetherUtil.resolveMavenArtifact(artId, groupId, version, extension, repos)
    }
    override fun resolveArtifact(artId: String, groupId: String, version: String, repos: jet.List<String>): File? {
        return AetherUtil.resolveMavenArtifact(artId, groupId, version, repos)
    }

    fun bootstrapGroupType(model: ContainerRoot, destGroupName: String, mservice: KevoreeModelHandlerService): AbstractGroupType? {
        //LOCATE NODE
        val optgroup = model.findGroupsByID(destGroupName)
        if(optgroup != null) {
            val groupTypeDeployUnit = optgroup.typeDefinition!!.deployUnit
            if (groupTypeDeployUnit!=null) {
                val kcl = installGroupType(optgroup.typeDefinition as GroupType)
                if (kcl != null) {
                    val clazz = kcl.loadClass(optgroup!!.typeDefinition!!.bean)
                    val groupType = clazz!!.newInstance() as AbstractGroupType

                    //ADD INSTANCE DICTIONARY
                    val dictionary: java.util.HashMap<String, Any> = java.util.HashMap<String, Any>()

                    val dictionaryType = optgroup!!.typeDefinition!!.dictionaryType

                    if (dictionaryType != null) {
                        for(dv in dictionaryType.defaultValues) {
                            dictionary.put(dv.attribute!!.name!!, dv.value!!)
                        }
                    }
                    val dictionaryModel = optgroup.dictionary
                    if (dictionaryModel != null) {
                        for(v in dictionaryModel.values) {
                            dictionary.put(v.attribute!!.name!!, v.value!!)
                        }
                    }
                    groupType.getDictionary()!!.putAll(dictionary)
                    groupType.setName(destGroupName)
                    groupType.setModelService(mservice)
                    groupType.setBootStrapperService(this)
                    return groupType
                } else {
                    Log.error("Unable to bootstrap Group so KCL pointer is null")
                    return null
                }
            } else {
                Log.error("NodeType deploy unit not found , have you forgotten to merge nodetype library ?")
                return null
            }
        }else {
            Log.error("Group not found using name " + destGroupName)
            return null
        }
    }

    private fun installGroupType(groupType: GroupType): Klassloader? {
        if (groupType.deployUnit != null ) {
            val superKCLs = ArrayList<Klassloader>()
            val superTypeBootStrap = groupType.superTypes.all {
                superType ->
                val superKCL = installGroupType(superType as org.kevoree.GroupType)
                if (superKCL != null) {
                    superKCLs.add(superKCL)
                }
                superKCL != null
            }
            //            val superTypeBootStrap = groupType.superTypes.all { superType -> installGroupTyp(superType as GroupType) != null }
            if (superTypeBootStrap) {
                var ct: DeployUnit? = null
                try {
                    ct = TypeDefinitionAspect().foundRelevantDeployUnit(groupType)
                } catch(e: Exception) {
                    e.printStackTrace()
                }
                if (ct != null) {
                    var kcl: Klassloader? = recursivelyInstallDeployUnit(ct!!)
                    if (kcl == null) {
                        Log.error("Unable to install the Node Type, maybe some libs cannot be installed")
                    } else {
                        superKCLs.forEach {
                            superKCL ->
                            (kcl as KevoreeJarClassLoader).addSubClassLoader(superKCL)
                        }
                    }
                    return kcl
                } else {
                    Log.error("Relevant Deploy Unit not found for " + groupType.name)
                    return null
                }
            } else {
                Log.error("Super installation failed")
                return null
            }
        } else {
            Log.error("The GroupType is not able to be executed on JavaSENode so the group cannot be instanciated. Please add a DeployUnit which enable the groupType execution on JavaSeNode.")
            return null
        }
    }

    private fun recursivelyInstallDeployUnit(du: DeployUnit): Klassloader? {
        val kcl_opt = installDeployUnit(du)

        var dpRes = true
        if (kcl_opt != null) {
            dpRes = du.requiredLibs.all { tp ->
                val idp = recursivelyInstallDeployUnit(tp)
                if (idp != null) {
                    kcl_opt.addSubClassLoader(idp)
                }
                idp != null
            }
        }
        if(kcl_opt != null && dpRes){
            return kcl_opt
        } else {
            Log.error("Unable to install {}:{}:{}", du.groupName, du.name, du.version)
            return null
        }
    }
}
